pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,196 @@
+  private transient int _numElements;
+
+  public VarLengthBytesValueReaderWriter(PinotDataBuffer dataBuffer) {
+    this._dataBuffer = dataBuffer;
+  }
+
+  public static long getRequiredSize(byte[][] byteArrays) {
+    // First include the magic header, length field and offsets array.
+    long length = INDEX_ARRAY_START_OFFSET + Integer.BYTES * (byteArrays.length);
+
+    for (byte[] array: byteArrays) {
+      length += array.length;
+    }
+    return length;
+  }
+
+  public static boolean isVarLengthBytesDictBuffer(PinotDataBuffer buffer) {
+    // If the buffer is smaller than header + numElements size, it's not var length dictionary.
+    if (buffer.size() > INDEX_ARRAY_START_OFFSET) {
+      byte[] header = new byte[MAGIC_HEADER_LENGTH];
+      buffer.copyTo(0, header, 0, MAGIC_HEADER_LENGTH);
+
+      if (Arrays.equals(MAGIC_HEADER, header)) {
+        // Also verify that there is a valid numElements value.
+        int numElements = buffer.getInt(MAGIC_HEADER_LENGTH);
+        return numElements > 0;
+      }
+    }
+
+    return false;
+  }
+
+  private void writeHeader() {
+    for (int offset = 0; offset < MAGIC_HEADER_LENGTH; offset++) {
+      _dataBuffer.putByte(offset, MAGIC_HEADER[offset]);
+    }
+  }
+
+  public void init(byte[][] byteArrays) {
+    this._numElements = byteArrays.length;
+
+    writeHeader();
+
+    // Add the number of elements as the first field in the buffer.
+    _dataBuffer.putInt(MAGIC_HEADER_LENGTH, byteArrays.length);
+
+    // Then write the offset of each of the byte array in the data buffer.
+    int nextOffset = INDEX_ARRAY_START_OFFSET;
+    int nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.putInt(nextOffset, nextArrayStartOffset);
+      nextOffset += Integer.BYTES;
+      nextArrayStartOffset += array.length;
+    }
+
+    // Finally write the byte arrays.
+    nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.readFrom(nextArrayStartOffset, array);
+      nextArrayStartOffset += array.length;
+    }
+  }
+
+  int getNumElements() {
+    // Lazily initialize the numElements.
+    if (_numElements == 0) {
+      _numElements = _dataBuffer.getInt(MAGIC_HEADER_LENGTH);
+      Preconditions.checkArgument(_numElements > 0);
+    }
+    return this._numElements;
+  }
+
+  @Override
+  public int getInt(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getLong(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public float getFloat(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public double getDouble(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
```

Comment:
Both getUnpaddedString() and getPaddedString() have the same code, and seem to return the same value. The callers of these api's might be expected different values.

Reply:
Correct and IMO, I think the `ValueReader` interface should only have `getString()` rather than padded and unpadded versions but because it has both of those methods, we can't pick/implement only one in VarLengthDictionary. 
Again, let me think more about a better way to clean this up. I wanted to keep the changes to minimal in my very first PR :)